staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,1236 @@
+                {
+                    warn!("TX: RTR frames not negotiated");
+                    return Err(Error::UnexpectedRtrFlag);
+                }
+                can_id |= flags | CAN_RTR_FLAG;
+            }
+
+            if (flags & VIRTIO_CAN_FLAGS_FD) != 0 {
+                if !self.check_features(VIRTIO_CAN_F_CAN_FD) {
+                    warn!("TX: FD frames not negotiated\n");
+                    return Err(Error::UnexpectedFdFlag);
+                }
+                flags |= CAN_FRMF_TYPE_FD;
+            } else {
+                if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC) {
+                    warn!("TX: Classic frames not negotiated\n");
+                    return Err(Error::UnexpectedClassicFlag);
+                }
+                flags = 0;
+            }
+
+            let corrected_request = VirtioCanFrame {
+                msg_type: msg_type.into(),
+                can_id: can_id.into(),
+                length: length.into(),
+                reserved: 0.into(),
+                flags: flags.into(),
+                sdu: request.sdu[0..64].try_into().unwrap(),
+            };
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1));
+            }
+
+            let response = match self.controller.write().unwrap().can_out(corrected_request) {
+                Ok(result) => result,
+                Err(_) => {
+                    warn!("We got an error from controller send func");
+                    VIRTIO_CAN_STATUS_ERR
+                }
+            };
+
+            desc_chain
+                .memory()
+                .write_slice(response.as_slice(), desc_response.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), desc_response.len())
+                .is_err()
+            {
+                trace!("Couldn't return used descriptors to the ring");
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    pub fn process_rx_requests(
+        &mut self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_rx_requests");
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        let desc_chain = &requests[0];
+        let descriptors: Vec<_> = desc_chain.clone().collect();
+
+        if descriptors.len() != 1 {
+            trace!("Error::UnexpectedDescriptorCount");
+            return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+        }
+
+        let desc_response = descriptors[0];
+        if !desc_response.is_write_only() {
+            return Err(Error::UnexpectedReadableDescriptor(1));
+        }
+
+        let mut response = match self.controller.write().unwrap().pop() {
```

Comment:
Maybe I misunderstood, but the rx queue is filled with empty frames when a guest driver requests frames. Getting a frame in the rx queue may not require an immediate response since the queue may be empty. Rather than treating an empty queue as an error, the request should be handled asynchronously when something arrives.

Reply:
When the guest triggers an `RX_QUEUE` notification the backend immidiately return `Ok(())` without doing anything.
```text
//vhu_can.rs:handle_event() function

if device_event == RX_QUEUE {
    trace!("RX_QUEUE\n");
    return Ok(());
};
```
So we only end up at the point where you pointed out when the `read_can_socket` function (running in a separate thread) reads a new CAN/FD message (from the input socket with name `can_in_name`), push the frame in the internal fifo queue `rx_fifo` (located in can.rs) and the trigger the `rx_event_fd` eventfd notification.

At that point, the handle_event() function will be called, recognises that the incoming event is the `BACKEND_EFD` (mmaped earlier with `rx_event_fd`), clean the eventfd `rx_event_fd` and then provide the correct queue (`vrings[RX_QUEUE]`) to the `process_rx_queue`.

Reply:
Oh, I see, so it is using the BACKEND_EFD event to notify the device that something has been enqueued and at that moment the requests at rx are handled by copying the frame from the queue to the request payload. For the moment, I can't see any issue of doing so but I need to think about it. I can only see an issue if the rx virtqueue is empty and a frame is received. This would trigger BACKEND_EFD, but the device would not be able to send it to the guest. In this case, resending will not be triggered if the guest fills the rx virtual queue.

Reply:
@stefano-garzarella pointed me that, to fix this, the rx virtqueue shall be processed also when the notification event is received.

Reply:
Ah ok! I would like to understand why though, can you elaborate more on this?
Based on what I find in QEMU's version of Opensynergy, the callback defined for `DEVICE_QUEUE_RX` queue index is NULL ([code reference](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L447)). Also, searching for `DEVICE_QUEUE_RX` value in `can_virtio.c` file, shows that this queue is accessed only by `virtio_can_receive` function ([here](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L214C16-L214C34)). Which as far as I have understood, is the one called each time a new CAN/FD frame is received by a certain CAN/FD host interface.

Reply:
With "RX queue is empty" I mean there are no buffers available for the device in the RX queue where to copy the incoming packet.

Reply:
Hello @stefano-garzarella and @MatiasVara, you were both right about it, the new version have fixed that issue.